README.md
# DatadogHttp

[![Module Version](https://img.shields.io/hexpm/v/datadog_http.svg)](https://hex.pm/packages/datadog_http)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/datadog_http/)
[![Total Download](https://img.shields.io/hexpm/dt/datadog_http.svg)](https://hex.pm/packages/datadog_http)
[![License](https://img.shields.io/hexpm/l/datadog_http.svg)](https://github.com/GhostPort/datadog_http_elixir/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/GhostPort/datadog_http_elixir.svg)](https://github.com/GhostPort/datadog_http_elixir/commits/master)
[![Coverage Status](https://coveralls.io/repos/github/GhostPort/datadog_http_elixir/badge.svg?branch=master)](https://coveralls.io/github/GhostPort/datadog_http_elixir?branch=master)

A client for Datadog's HTTP API, useful for building integrations.

_This community library is not an official product or project of Datadog. Maintainers are not affiliated with Datadog in any way._

### Supported endpoints

- [ ] Audit
- [ ] Authentication
- [ ] AuthN Mappings
- [ ] AWS Integration
- [ ] AWS Logs Integration
- [ ] Azure Integration
- [ ] Cloud Workload Security
- [ ] Dashboard Lists
- [ ] Dashboards
- [ ] Downtimes
- [ ] Embeddable Graphs
- [ ] Events
  - [ ] Get a list of events
  - [x] Post an event
  - [ ] Get an event
  - [ ] Search events
- [ ] GCP Integration
- [ ] Hosts
- [ ] Incident Services
- [ ] Incident Teams
- [ ] Incidents
- [ ] IP Ranges
- [ ] Key Management
- [ ] Logs
- [ ] Logs Archives
- [ ] Logs Indexes
- [ ] Logs Metrics
- [ ] Logs Pipelines
- [ ] Logs Restriction Queries
- [ ] Metrics
  - [ ] Create a tag configuration
  - [ ] Get active metrics list
  - [ ] Submit distribution points
  - [x] Submit metrics
  - [ ] Get metric metadata
  - [ ] List tag configuration by name
  - [ ] Edit metric metadata
  - [ ] Update a tag configuration
  - [ ] Delete a tag configuration
  - [ ] Search metrics
  - [ ] Get a list of metrics
  - [ ] Query timeseries points
  - [ ] List tags by metric name
  - [ ] List active tags and aggregations
  - [ ] List distinct metric volumes by metric name
  - [ ] Configure tags for multiple metrics
  - [ ] Tag Configuration Cardinality Estimator
- [ ] Monitors
- [ ] Notebooks
- [ ] Opsgenie Integration
- [ ] Organizations
- [ ] PagerDuty Integration
- [ ] Processes
- [ ] Roles
- [ ] RUM
- [ ] Screenboards
- [ ] Security Monitoring
- [ ] Service Accounts
- [ ] Service Checks
- [ ] Service Dependencies
- [ ] Service Level Objective Corrections
- [ ] Service Level Objectives
- [ ] Slack Integration
- [ ] Snapshots
- [ ] Synthetics
- [ ] Tags
- [ ] Timeboards
- [ ] Usage Metering
- [ ] Users
- [ ] Webhooks Integration

## Installation

Add to your dependencies in `mix.exs`. The hex specification is required.

```elixir
def deps do
  [
    {:datadog_http, "~> 0.0.1"}
  ]
end
```

## Configuration

All calls to Datadog require an API key. Add the following configuration
to your project to set the values. This configuration is optional, see below for a
runtime configuration. The library will raise an error if the relevant credentials
are not provided either via `config.exs` or at runtime.

```elixir
config :datadog_http,
  base_url: "https://api.datadoghq.com",
  api_key: "your_client_id",
  adapter: Tesla.Adapter.Hackney, # optional
  http_options: [timeout: 10_000, recv_timeout: 30_000] # optional
```

By default, `base_url` is set to the main datadog API endpoint (`https://api.datadoghq.com`).

## Runtime configuration

Alternatively, you can provide the configuration at runtime. The configuration passed
as a function argument will overwrite the configuration in `config.exs`, if one exists.

For example, if you want to hit a different URL when calling the metrics submission endpoint, you could
pass in a configuration argument to `DatadogHttp.Metrics.submit/2`.

```elixir
DatadogHttp.Metrics.submit(
  [],
  %{base_url: "https://api.datadoghq.eu", api_key: "an-api-key"}
)
```


CHANGELOG.md
# Changelog

## 0.0.3 - 2022-10-07

### Added

- Events can now be posted to Datadog.

### Changed

- Fixed issue where the `Point` struct name was sent to the metrics endpoint.

## 0.0.2 - 2022-10-07

### Added

- `DateTime` instances are now accepted in the `Point` structure of metrics. The existing integer Unix timestamp option remains.

## 0.0.1 - 2022-10-06

**Initial release**

### Added

- Basic structure for the API client.
- Metrics submit method.


test/lib/datadog_http/client_test.exs
defmodule DatadogHttp.ClientTest do
  use ExUnit.Case

  alias DatadogHttp.Client

  @moduletag :client

  describe "client new/1" do
    @describetag :unit

    setup do
      on_exit(fn ->
        Application.delete_env(:datadog_http, :api_key)
      end)
    end

    test "base url" do
      # raise when base URL is nil
      Application.put_env(:datadog_http, :base_url, nil)
      Application.put_env(:datadog_http, :api_key, "irrelevant-api-key")

      assert_raise DatadogHttp.MissingBaseUrlError, fn ->
        Client.new()
      end

      # use the application environment variable
      Application.put_env(:datadog_http, :base_url, "https://environment-base-url")

      client = Client.new()

      assert Enum.member?(
               client.pre,
               {Tesla.Middleware.BaseUrl, :call, ["https://environment-base-url/api"]}
             )

      # runtime config overrides app env
      client = Client.new(%{base_url: "https://config-base-url"})

      assert Enum.member?(
               client.pre,
               {Tesla.Middleware.BaseUrl, :call, ["https://config-base-url/api"]}
             )
    end

    test "API key" do
      # raise when API key is not provided
      assert_raise DatadogHttp.MissingAPIKeyError, fn ->
        Client.new()
      end

      # use the application environment variable
      Application.put_env(:datadog_http, :api_key, "env-api-key")

      client = Client.new()

      assert Enum.find(client.pre, fn
               {Tesla.Middleware.Headers, _, [headers]} ->
                 Enum.member?(headers, {"DD-API-KEY", "env-api-key"})

               _ ->
                 false
             end),
             "API key should match the application environment variable value"

      # runtime config overrides app env
      Application.put_env(:datadog_http, :api_key, "env-api-key")

      client = Client.new(%{api_key: "config-api-key"})

      assert Enum.find(client.pre, fn
               {Tesla.Middleware.Headers, _, [headers]} ->
                 Enum.member?(headers, {"DD-API-KEY", "config-api-key"})

               _ ->
                 false
             end),
             "API key should match the configuration dictionary value"
    end
  end
end


test/lib/datadog_http/events_test.exs
defmodule DatadogHttp.EventsTest do
  use ExUnit.Case

  alias DatadogHttp.Events

  @moduletag :events

  setup do
    bypass = Bypass.open()
    config = %{api_key: "some_key", base_url: "http://localhost:#{bypass.port}"}

    %{bypass: bypass, config: config}
  end

  describe "post" do
    @describetag :unit

    test "documentation example success", %{bypass: bypass, config: config} do
      event = %Events.Event{
        title: "Example-Post_an_event_returns_OK_response",
        text: "A text message.",
        tags: [
          "test:ExamplePostaneventreturnsOKresponse"
        ]
      }

      response_event = Map.merge(event, %{id: 123, id_str: "123"})

      Bypass.expect(bypass, "POST", "/api/v1/events", fn conn ->
        {:ok, body, conn} = Plug.Conn.read_body(conn)

        body = Jason.decode!(body, keys: :atoms)

        assert %{
                 title: "Example-Post_an_event_returns_OK_response",
                 text: "A text message.",
                 tags: [
                   "test:ExamplePostaneventreturnsOKresponse"
                 ]
               } = body

        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(202, Jason.encode!(%{event: response_event}))
      end)

      assert {:ok, ^response_event} = Events.post(event, config)
    end

    test "Datetime and atom success", %{bypass: bypass, config: config} do
      event = %Events.Event{
        title: "Example-Post_an_event_returns_OK_response",
        text: "A text message.",
        date_happened: ~U[2022-01-01 00:00:00Z],
        alert_type: :error,
        priority: :low
      }

      response_event = Map.merge(event, %{id: 123, id_str: "123"})

      Bypass.expect(bypass, "POST", "/api/v1/events", fn conn ->
        {:ok, body, conn} = Plug.Conn.read_body(conn)

        body = Jason.decode!(body, keys: :atoms)

        assert %{
                 date_happened: 1_640_995_200,
                 alert_type: "error",
                 priority: "low"
               } = body

        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(202, Jason.encode!(%{event: response_event}))
      end)

      assert {:ok, ^response_event} = Events.post(event, config)
    end

    test "error response", %{bypass: bypass, config: config} do
      event = %Events.Event{}

      Bypass.expect(bypass, "POST", "/api/v1/events", fn conn ->
        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(400, Jason.encode!(%{errors: ["Bad Request"]}))
      end)

      assert {:error, ["Bad Request"]} = Events.post(event, config)
    end

    test "error" do
      event = %Events.Event{}
      config = %{api_key: "some_key", base_url: "nope"}

      assert {:error, :nxdomain} = Events.post(event, config)
    end

    test "malformed_response", %{bypass: bypass, config: config} do
      event = %Events.Event{}

      Bypass.expect(bypass, "POST", "/api/v1/events", fn conn ->
        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(200, Jason.encode!(%{bad: "response"}))
      end)

      assert :error = Events.post(event, config)
    end
  end
end


test/lib/datadog_http/metrics_test.exs
defmodule DatadogHttp.MetricsTest do
  use ExUnit.Case

  alias DatadogHttp.Metrics

  @moduletag :metrics

  setup do
    bypass = Bypass.open()
    config = %{api_key: "some_key", base_url: "http://localhost:#{bypass.port}"}

    %{bypass: bypass, config: config}
  end

  describe "submit" do
    @describetag :unit

    test "no metrics success", %{bypass: bypass, config: config} do
      Bypass.expect(bypass, "POST", "/api/v2/series", fn conn ->
        {:ok, body, conn} = Plug.Conn.read_body(conn)

        assert %{series: []} = Jason.decode!(body, keys: :atoms)

        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(202, Jason.encode!(%{errors: []}))
      end)

      assert :ok = Metrics.submit([], config)
    end

    test "documentation example success", %{bypass: bypass, config: config} do
      metric = %Metrics.Metric{
        metric: "system.load1",
        type: 0,
        points: [
          %Metrics.Point{
            timestamp: 1_636_629_071,
            value: 0.7
          }
        ],
        resources: [
          %Metrics.Resource{
            name: "dummyhost",
            type: "host"
          }
        ]
      }

      Bypass.expect(bypass, "POST", "/api/v2/series", fn conn ->
        {:ok, body, conn} = Plug.Conn.read_body(conn)

        body = Jason.decode!(body, keys: :atoms)

        assert length(body.series) == 1

        assert %{
                 metric: "system.load1",
                 type: 0,
                 points: [
                   %{
                     timestamp: 1_636_629_071,
                     value: 0.7
                   }
                 ],
                 resources: [
                   %{
                     name: "dummyhost",
                     type: "host"
                   }
                 ]
               } = Enum.at(body.series, 0)

        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(202, Jason.encode!(%{errors: []}))
      end)

      assert :ok = Metrics.submit([metric], config)
    end

    test "DateTime objects are converted into unix second timestamps", %{
      bypass: bypass,
      config: config
    } do
      metric = %Metrics.Metric{
        metric: "something.measured",
        points: [
          %Metrics.Point{
            timestamp: ~U[2022-01-01 00:00:00Z],
            value: 4.2
          }
        ]
      }

      Bypass.expect(bypass, "POST", "/api/v2/series", fn conn ->
        {:ok, body, conn} = Plug.Conn.read_body(conn)

        body = Jason.decode!(body, keys: :atoms)

        assert length(body.series) == 1

        assert %{
                 metric: "something.measured",
                 points: [
                   %{
                     timestamp: 1_640_995_200,
                     value: 4.2
                   }
                 ]
               } = Enum.at(body.series, 0)

        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(202, Jason.encode!(%{errors: []}))
      end)

      assert :ok = Metrics.submit([metric], config)
    end

    test "error response", %{bypass: bypass, config: config} do
      Bypass.expect(bypass, "POST", "/api/v2/series", fn conn ->
        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(400, Jason.encode!(%{errors: ["Bad Request"]}))
      end)

      assert {:error, ["Bad Request"]} = Metrics.submit([], config)
    end

    test "error" do
      config = %{api_key: "some_key", base_url: "nope"}

      assert {:error, :nxdomain} = Metrics.submit([], config)
    end

    test "malformed_response", %{bypass: bypass, config: config} do
      Bypass.expect(bypass, "POST", "/api/v2/series", fn conn ->
        conn
        |> Plug.Conn.put_resp_header("Content-Type", "application/json")
        |> Plug.Conn.resp(200, Jason.encode!(%{bad: "response"}))
      end)

      assert :error = Metrics.submit([], config)
    end
  end
end


mix.exs
defmodule DatadogHttp.MixProject do
  use Mix.Project

  @version "0.0.3"

  def project do
    [
      app: :datadog_http,
      version: @version,
      elixir: "~> 1.14",
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test, "coveralls.html": :test, "coveralls.github": :test],
      name: "Datadog HTTP",
      source_url: "https://github.com/GhostPort/datadog_http_elixir"
    ]
  end

  def application do
    [
      extra_applications: [:logger],
      env: env()
    ]
  end

  defp env do
    [
      base_url: "https://api.datadoghq.com"
    ]
  end

  defp deps do
    [
      {:tesla, "~> 1.4"},
      {:hackney, "~> 1.18"},
      {:jason, "~> 1.4"},
      {:exconstructor, "~> 1.2"},
      {:ex_doc, "~> 0.23", only: :dev, runtime: false},
      {:bypass, "~> 2.1", only: :test},
      {:excoveralls, "~> 0.15", only: :test}
    ]
  end

  defp description do
    "A client for Datadog's HTTP API, useful for building integrations."
  end

  defp package do
    [
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/GhostPort/datadog_http_elixir",
        "Datadog API Reference" => "https://docs.datadoghq.com/api/latest/"
      },
      maintainers: ["Christian Alexander"]
    ]
  end
end


lib/datadog_http.ex
defmodule DatadogHttp do
  @moduledoc """
  A client for Datadog's HTTP API, useful for building integrations.

  [Datadog API reference](https://docs.datadoghq.com/api/latest/)
  """

  defmodule MissingAPIKeyError do
    defexception message: """
                 An `api_key` is required for calls to Datadog.
                 Configure `api_key` in your config.exs file or pass it into the function via the `config` argument.

                 config :datadog_http, api_key: "your_api_key"
                 """
  end

  defmodule MissingBaseUrlError do
    defexception message: """
                 The `base_url` is required for calls to Datadog.
                 Configure `base_url` in your config.exs file or pass it into the function via the `config` argument.

                 By default, this is set to "https://api.datadoghq.com".

                 For help finding your Datadog base URL subdomain, check out the related Datadog documentation (https://docs.datadoghq.com/getting_started/site/#access-the-datadog-site)

                 config :datadog_http, base_url: "https://api.datadoghq.eu"
                 """
  end
end


lib/datadog_http/client.ex
defmodule DatadogHttp.Client do
  @moduledoc """
  Functions to build a Tesla client for handling HTTP requests.
  """

  @doc """
  Creates a new Tesla client.
  Optional `config` argument sets the following values at runtime:
  - Base URL
  - API Key
  - `Tesla.Adapter`
  - `Tesla.Middleware`
  - HTTP options for selected `Tesla.Adapter`
  Example
  ```
  config = %{
    base_url: "https://api.datadoghq.com",
    api_key: "some-api-key",
    adapter: Tesla.Adapter.Httpc,
    middleware: [Tesla.Middleware.Logger],
    http_options: [recv_timeout: 10_000]
  }
  client = Client.new(config)
  ```
  """
  @spec new(map) :: Tesla.Client.t()
  def new(config \\ %{}) do
    middleware = [
      {Tesla.Middleware.BaseUrl, get_base_url(config) <> "/api"},
      {Tesla.Middleware.Headers,
       [
         {"Content-Type", "application/json"},
         {"DD-API-KEY", get_api_key(config)}
       ]},
      Tesla.Middleware.JSON
    ]

    adapter = {get_adapter(config), get_http_options(config)}

    Tesla.client(middleware, adapter)
  end

  defp get_base_url(config) do
    case config[:base_url] || Application.get_env(:datadog_http, :base_url) do
      nil ->
        raise DatadogHttp.MissingBaseUrlError

      base_url ->
        base_url
    end
  end

  defp get_api_key(config) do
    case config[:api_key] || Application.get_env(:datadog_http, :api_key) do
      nil ->
        raise DatadogHttp.MissingAPIKeyError

      api_key ->
        api_key
    end
  end

  defp get_adapter(config) do
    config[:adapter] || Application.get_env(:datadog_http, :adapter) || Tesla.Adapter.Hackney
  end

  defp get_http_options(config) do
    Keyword.merge(
      Application.get_env(:datadog_http, :http_options, []),
      config[:http_options] || []
    )
  end
end


lib/datadog_http/metrics.ex
defmodule DatadogHttp.Metrics do
  @moduledoc """
  Functions for Datadog metrics endpoints.
  """

  alias DatadogHttp.Client

  @type config :: %{required(atom) => String.t() | keyword}
  @type error :: {:error, [String.t()]} | {:error, any()} | :error

  defmodule Metric do
    @moduledoc """
    Datadog metric data structure.
    """

    @derive Jason.Encoder
    defstruct interval: nil,
              metadata: nil,
              metric: nil,
              points: [],
              resources: [],
              source_type_name: nil,
              tags: [],
              type: nil,
              unit: nil

    @type t :: %__MODULE__{
            interval: integer(),
            metadata: map(),
            metric: String.t(),
            points: [DatadogHttp.Metrics.Point],
            resources: [DatadogHttp.Metrics.Resource],
            source_type_name: String.t(),
            tags: [String.t()],
            type: String.t(),
            unit: String.t()
          }
  end

  defmodule Point do
    @moduledoc """
    Datadog metric point data structure.
    """

    defstruct timestamp: nil,
              value: nil

    @type t :: %__MODULE__{
            timestamp: integer() | DateTime,
            value: float()
          }

    defimpl Jason.Encoder, for: __MODULE__ do
      def encode(%{timestamp: %DateTime{} = timestamp} = value, opts) do
        __MODULE__.encode(%{value | timestamp: DateTime.to_unix(timestamp)}, opts)
      end

      def encode(value, opts) do
        value
        |> Map.delete(:__struct__)
        |> Jason.Encode.map(opts)
      end
    end
  end

  defmodule Resource do
    @moduledoc """
    Datadog metric resource data structure.
    """

    @derive Jason.Encoder
    defstruct name: nil,
              type: nil

    @type t :: %__MODULE__{
            name: String.t(),
            type: String.t()
          }
  end

  @spec submit([Metric]) :: :ok | error
  def submit(metrics, config \\ %{}) do
    c = config[:client] || Client.new(config)

    case Tesla.post(c, "/v2/series", %{series: metrics}) do
      {:ok, %{status: status, body: %{"errors" => errors}}}
      when status >= 400 and is_list(errors) ->
        {:error, errors}

      {:ok, %{status: 202}} ->
        :ok

      {:error, reason} ->
        {:error, reason}

      _ ->
        :error
    end
  end
end


lib/datadog_http/events.ex
defmodule DatadogHttp.Events do
  @moduledoc """
  Functions for Datadog events endpoints.
  """

  alias DatadogHttp.Client

  @type config :: %{required(atom) => String.t() | keyword}
  @type error :: {:error, [String.t()]} | {:error, any()} | :error

  defmodule Event do
    @moduledoc """
    Datadog event data structure.
    """

    @type alert_type ::
            :error | :warning | :info | :success | :user_update | :recommendation | :snapshot
    @type event_priority :: :normal | :low

    defstruct [
      :text,
      :title,
      id: nil,
      id_str: nil,
      aggregation_key: nil,
      alert_type: nil,
      date_happened: nil,
      device_name: nil,
      host: nil,
      priority: nil,
      related_event_id: nil,
      source_type_name: nil,
      tags: []
    ]

    use ExConstructor

    @type t :: %__MODULE__{
            aggregation_key: String.t(),
            alert_type: alert_type,
            date_happened: integer | DateTime,
            device_name: String.t(),
            host: String.t(),
            id: integer(),
            id_str: String.t(),
            priority: event_priority,
            related_event_id: integer,
            source_type_name: String.t(),
            tags: [String.t()],
            text: String.t(),
            title: String.t()
          }

    defimpl Jason.Encoder, for: __MODULE__ do
      def encode(%{date_happened: %DateTime{} = date_happened} = value, opts) do
        __MODULE__.encode(%{value | date_happened: DateTime.to_unix(date_happened)}, opts)
      end

      def encode(value, opts) do
        alert_type_string = if value.alert_type, do: Atom.to_string(value.alert_type)
        priority_string = if value.priority, do: Atom.to_string(value.priority)

        value
        |> Map.merge(%{
          alert_type: alert_type_string,
          priority: priority_string
        })
        |> Map.delete(:__struct__)
        |> Jason.Encode.map(opts)
      end
    end
  end

  defp map_event(body_event) do
    alert_type =
      if body_event["alert_type"],
        do: String.to_existing_atom(body_event["alert_type"]),
        else: nil

    priority =
      if body_event["priority"], do: String.to_existing_atom(body_event["priority"]), else: nil

    date_happened =
      if body_event["date_happened"],
        do: DateTime.from_unix!(body_event["date_happened"]),
        else: nil

    Event.new(body_event)
    |> Map.merge(%{
      alert_type: alert_type,
      priority: priority,
      date_happened: date_happened
    })
  end

  @spec post(Event) :: {:ok, Event.t()} | error
  def post(event, config \\ %{}) do
    c = config[:client] || Client.new(config)

    case Tesla.post(c, "/v1/events", event) do
      {:ok, %{status: status, body: %{"errors" => errors}}}
      when status >= 400 and is_list(errors) ->
        {:error, errors}

      {:ok, %{status: 202, body: %{"event" => event}}} ->
        {:ok, map_event(event)}

      {:error, reason} ->
        {:error, reason}

      _ ->
        :error
    end
  end
end
